mlflow/server/js/src/Actions.test.js

```diff
@@ -1,5 +1,54 @@
-import { ErrorWrapper } from './Actions';
+import { ErrorWrapper, wrapDeferred } from './Actions';
+
+/**
+ * Returns mock Ajax function that sequentially responds with status codes in `responseStatusCodes`,
+ * looping back to the start of the array of status codes if necessary.
+ */
+const getMockAjax = (responseStatusCodes) => {
+  let i = 0;
+  return ({success, error}) => {
+    const statusCode = responseStatusCodes[i % responseStatusCodes.length];
+    if (statusCode === 200) {
+      success({requestIndex: i});
+    } else {
+      error({status: statusCode});
+    }
+    i += 1;
+  };
+};
 
 test("ErrorWrapper.getErrorCode does not fail on JSON decoding problems", () => {
   new ErrorWrapper({ responseText: 'a{waefaw' });
 });
+
+test('wrapDeferred retries on 429s', (done) => {
+  wrapDeferred(getMockAjax([429, 429, 200]), {}, 1000, 10).then((result) => {
+    expect(result).toEqual({requestIndex: 2});
+  }).then(() => done());
+});
+
+test('wrapDeferred responds with 429 after timeout period', (done) => {
+  // First request responds with 429, we then retry and receive another 429, but don't have time to
+  // retry again, so we propagate the 429
+  let caughtError = false;
+  wrapDeferred(getMockAjax([429, 429, 200]), {}, 10, 10).catch((result) => {
```

Comment:
Maybe try to see if this work with 600, 10 instead of 10, 10?

Reply:
Ah do you mean 10, 600? (max wait interval of 10 ms but a larger sleep interval, e.g. 600 ms?) Here we actually do want to exhaust the max wait interval. Manually verified that things still work with 10, 600 :)